cpp/ql/src/semmle/code/cpp/File.qll

```diff
@@ -302,7 +302,13 @@ class File extends Container, @file {
   predicate compiledAsMicrosoft() {
     exists(Compilation c |
       c.getAFileCompiled() = this and
-      c.getAnArgument() = "--microsoft"
+      (
+        c.getAnArgument() = "--microsoft" or
+        c.getAnArgument().matches("%\\\\cl.exe")
+      )
```

Comment:
We discussed the matter on Slack a few days ago, and agreed that the right way would be for the extractor to tell us explicitly whether a compilation was Microsoft or not.

Right now we might get slightly better accuracy by looking for the existence of the `_MSC_VER` macro anywhere in the snapshot, though we'll just have to assume that all files are compiled as Microsoft if we see it (so we'll do worse in the probably very rare case of mixed Microsoft and non-Microsoft compilations).  Another suggestion was looking for paths beginning `C:` or similar (which detects a Microsoft file system, rather than compiler, and may work poorly with test path normalization).

I'm happy to implement the `_MSC_VER` thing if you're convinced it would be preferable.  Or just make the `cl.exe` clause a bit more robust?

Reply:
In that case, it sounds better to make the `cl.exe` test more robust. Is it only relevant to test for `cl.exe` when there's also a `--mimic` argument?

Reply:
Yes I think so.  My logic was that with matching the `\`, it's highly unlikely we'll get false results for `cl.exe`.

Reply:
A quick Google search shows that some people like to spell it `CL.exe`, so turning the match into a case-insensitive regex might be necessary. What `Compilation` argument do we get if the compiler is just invoked as `cl`, without the `.exe`? Or the `--mimic` option only work if `.exe` is included? I suppose this argument is only ever produced by our own tracer, so your current match might be fine if the tracer normalises everything.